app/services/session.js

```diff
@@ -70,24 +70,49 @@ export default class SessionService extends Service {
     // the popup window prevention mechanism of the browser, since the async opening
     // can not be associated with the original user click event
     let { url } = await ajax(`/api/private/session/begin`);
+    let expectedState = new URLSearchParams(new URL(url).search).get('state');
     win.location = url;
 
-    let event = await race([waitForEvent(window, 'message'), this.windowCloseWatcherTask.perform(win)]);
-    if (event.closed) {
-      this.notifications.warning('Login was canceled because the popup window was closed.');
-      return;
-    }
+    let code, state;
+    const windowClosed = this.windowCloseWatcherTask.perform(win);
 
-    win.close();
-    if (event.origin !== window.location.origin || !event.data) {
-      return;
-    }
+    while (!code) {
+      const event = await race([waitForEvent(window, 'message'), windowClosed]);
```

Comment:
instead of having the `while` loop in this task, I'm wondering if it makes sense to extract a `stateMessageWatcherTask` sort of thing, that we can call instead of the `waitForEvent()`. That task would then call `waitForEvent()` internally in a loop until it sees the message that we're waiting for.

Reply:
I've adjusted the code to extract the child task. let's see if that fixes it :)